### ISSUE TITLE
process for evolving api

### ISSUE THREAD

**Issue by lbergelson** [samtools/htsjdk#378]
I think we should formalize a process for evolving the api and removing deprecated functions.  We have some things that have been deprecated for years and never removed.  

I'd recommend something like "functions that are deprecated are subject to removal after n months."  Where N is something like 3-6.  We'd commit to documenting what was removed and how to replace it if relevant in those releases.

**Comment by yfarjoun** [comment]
I agree with the sprit, but I think that 3-6 months is too fast. We already
have users (eg igv) who inform us of the trouble they go through when they
rev htsjdk, removing deprecated functions after 3 months would make it much
harder for users like that to use the library. I think that something on
the order of 1 year, assuming there was a discussion of the (specific)
deprecation as an issue, and an agreement among the users would be more
appropriate...but that's just my 2ct.

Y

**Comment by tfenne** [comment]
I tend to agree with @yfarjoun - deprecated things should have a fairly long twilight period before being yanked.  While high volume users of the library and likely to update frequently, I suspect there is a fairly long tail of users who don't update all that often and would get bitten by that kind of change.  I would think something in the 9-12 month range would be more realistic.

**Comment by lbergelson** [comment]
I think it would be a good idea to pair a deprecation policy with a 'experimental' or 'beta' tag policy, so that new functionality wouldn't be locked in immediately.  

Basically: new functionality that isn't quite solid gets tagged with `@beta` or `@experimental`.

We reserve the rights to change these methods at will.  That way we can experiment with new ideas without being immediately committed to supporting them forever.  At some point, experimental methods lose their experimental status.  Then we've committed to not changing them until they've been deprecated.

Three months is probably fast.  A year seems slow to me.  Whatever the number is isn't that important, it's just important that we choose a number and can then act to remove things at some point.

**Comment by vadimzalunin** [comment]
I would advise to separate API and implementation even further. APIs are generally much more stable and that's ok. At the moment it looks like we tend to classify all public classes/methods as API but they are not API imo. Ideally an API would have a version, which is different from htsjdk version.

**Comment by cmnbroad** [comment]
I've added a new  "version" label to htsjdk to use with PRs or issues that affect the public API in a way that requires a version number change. I've also [tagged](https://github.com/samtools/htsjdk/issues?utf8=%E2%9C%93&q=label%3Aversion) the PRs that I know of since the 2.0.1 release. I think it would be useful if going forward, submitters and reviewers would add that label to any PRs or issues that affect api compatibility. It will help to determine how to bump the new semantic version # when we go to release.

Sine 2.0.1, we added the javadoc @deprecated tag to a method in SAMFileValidator (not sure if this was deliberate but interestingly the method itself isn't deprecated). Still, it breaks the hellbender build since it treats warnings as errors. How should that affect the version number? We also changed/moved some public CRAM APIs. The latter seems to warrant a major version number change. However, I expect that we will continue to refactor the CRAM code in way that will require public API changes in future versions.

**Comment by droazen** [comment]
@cmnbroad After some discussion, we agreed that this release will be tagged as 2.1.0, and we'll add a note like this to the project README: "htsjdk is migrating to semantic versioning (link to semantic versioning definition). We will eventually adhere to it strictly and bump our major version whenever there are breaking changes to our API, but until we more clearly define what constitutes our official API, clients should assume that every release potentially contains at least minor changes to public methods."

**Comment by akiezun** [comment]
The definition of API needs to be carefully considered. Technically, because most htsjdk classes are not `final` (and thus subclassing is allowed and subclasses are considered API customers) pretty much *everything* is public API - ie even the order of operations in a private method can affect subclasses. In this trivial example, `SuperA` cannot reorder `m1` and `m2` in the private method `m` because it will mess up the world for `SubA`

```
class SuperA{
  public void a(){ m(); } 
  private void m(){ m1(); m2();}
  public void m1(){ /*SuperA does not care about the order*/ } 
  public void m2(){ /*SuperA does not care about the order*/ }
}

class SubA extends SuperA {
  public void m1() { System.out.println("I rely on this being called first"); }
  public void m2() { System.out.println("I rely on this being called second"); }
}

SuperA a = new SubA();
a.m();
```

**Comment by tfenne** [comment]
I've been spending some time lately writing some code in [rust](http://www.rust-lang.org) and they have an interesting idea that might be worth bringing up here.  On top of the usual mechanisms (access levels, final/not-final) they have a notion of stable/unstable APIs.  In rust this is managed by the compiler and macros, but it could easily be done using annotations in Java.  

The idea is that public classes and functions can be marked as `unstable`.  The contract is then that anything marked `unstable` is not held to the same standard with respect to versioning.  This gives people a choice - stick to APIs that have been stabilized and you can rely on backwards-compatibility inline with `SemVer` version numbers, or dip into the `unstable` and be prepared to deal with compatibility breaks.  It allows things to be public and available for experimentation without requiring that be locked down, or bumping major version numbers left and right.

We could go through a fairly rapid process and decide what of the public/protected stuff in htsjdk is stable/unstable, being quite conservative, and then work towards "stabilizing" more of the API.

**Comment by lbergelson** [comment]
@tfenne  I've been thinking about ways to manage this as well.  

The spark / hadoop projects use what you're describing, annotations marking the stability level of classes and methods.  ([some annotations here](https://github.com/apache/spark/tree/master/core/src/main/java/org/apache/spark/annotation))  

The key element we need is tooling to support automatically identifying changes to the api.  Otherwise we are going to have a very difficult time accurately stating when we change the api.  I'm sure the spark project has something automated to check this, but I'm not sure what they're using.  

There are a few tools that exist to check binary compatibility between two jars [japicmp](https://github.com/siom79/japicmp) and [revapi](http://revapi.org/) are two.  They can have exclusion lists, but I think those work by package/classname, I'm not sure its possible to mark one method as unstable in an otherwise stable class for instance.  

Another (much more invasive) approach I've seen is to physically separate the API and implementation into separate jars.  This is likely impractical though since HTSJDK isn't separable into interfaces and implementing classes and doing so would be a major project.  (Maybe a good one to do at some point.  It's really a shame that `SAMRecord` isn't an implementation of a `Read` interface.)  It would reduce the need for separate tooling to do it this way though, since it would be more obvious when the API version needed to change.
